### ISSUE TITLE
Policy for updating ICU in a release line

### ISSUE THREAD

**Issue by richardlau** [nodejs/Release#679]
# Background

Started from https://github.com/nodejs/node/pull/38178#issuecomment-863116223 -- Node.js 14.17.1 was released with a V8 change that breaks compiling against a system ICU 65. We specify a minimum version of ICU in `tools/icu/icu_versions.json` and this was:
- 65 in Node.js 14.0.0
- 67 in Node.js 14.6.0 as part of the V8 8.4 update https://github.com/nodejs/node/pull/34356

14.17.1 included https://github.com/nodejs/node/pull/38497 to update to ICU 69 but also included a V8 backport (https://github.com/nodejs/node/commit/40df0dcb59c6769eb013d40dc0ce5de66210f3bd) and it's this backport which has broken the compilation with ICU 65 (and probably 66 going by the commit message from the original V8 commit).

We had a brief chat in today's Release WG meeting (which was recorded (link tbc)) and the agreed action was that @richardlau would see if its possible to patch V8 to restore being able to compile against ICU 65 while also still being able to compile against ICU 69. If it's not (or proves too complicated) then we'll revert the ICU 69 update in 14.x.

# Policy discussion

While we have a plan for the immediate issue on 14.x in the meeting we wanted to continue to the generalized policy discussion around ICU, which will continue in this issue. 

I'd like to advocate a "we won't change the minimum ICU during a major" policy -- there are some considerations.

The minimum ICU is a combination of:
- What is required by Node.js (specified in [`tools/icu/icu_versions.json`](https://github.com/nodejs/node/blob/6c77fd3a5359b16d472c23cd1be2c208b7bcb1d2/tools/icu/icu_versions.json#L2))
- What is required by V8 (specified in [`deps/v8/src/objects/intl-objects.h`](https://github.com/nodejs/node/blob/6c77fd3a5359b16d472c23cd1be2c208b7bcb1d2/deps/v8/src/objects/intl-objects.h#L24))

We have a [test](https://github.com/nodejs/node/blob/6c77fd3a5359b16d472c23cd1be2c208b7bcb1d2/test/parallel/test-icu-minimum-version.js) to ensure consistency for the declared minimum versions. V8 backports complicate things (like in the 14.x case) as they may come from upstream V8 releases where the minimum has been moved on and may introduce incompatible code. 

Fixing the minimum ICU version may prevent V8 updates from landing, for example @targos mentioned that the V8 9.2 update will also bump the minimum ICU to 69. If we enact a "no minimum version bump" that might prevent updating Node.js 16 to V8 9.2.

Another possibility/addition to fixing the minimum ICU was to not update ICU at all in a release line. Are there any potential downsides? We'd still be able to update timezone information (documentation on how to do so added in https://github.com/nodejs/node/pull/30364).

For completion there was also some previous discussion relating to not updating ICU in a release line related to the full-icu package (https://github.com/nodejs/Release/issues/483) -- that should be less of a concern for official releases of Node.js 14 onwards as those are compiled with `full-icu` and don't require the external `full-icu` npm.

cc @nodejs/releasers @AdamMajer @nodejs/i18n-api

**Comment by srl295** [comment]
> V8 backports complicate things

yes, they tend to use the very latest ICU stable APIs and sometimes even draft.  There is _sometimes_ a different API that could be used which was already available (i.e. with a less convenient API) with equivalent functionality, but more often than not there are functional changes which are needed as well.

Not changing the minimum ICU during a major is a good general policy.  What isn't discussed here is the reason for using a different bundled ICU version:
- platform ICU - your OS already has an ICU and you want to use that.  example: linux distro, or even a large application environment bundling node that also has an existing ICU.    homebrew might be in this category
- custom built ICU - you have a custom version perhaps with custom data.

**Comment by srl295** [comment]
it should also be noted that ICU is called directly by node.js in some places such as encoders.

**Comment by mhdawson** [comment]
+1 to Not changing the minimum ICU during a major seems like a good general policy.

**Comment by targos** [comment]
> I'd like to advocate a "we won't change the minimum ICU during a major" policy -- there are some considerations.

+1

> Fixing the minimum ICU version may prevent V8 updates from landing, for example @targos mentioned that the V8 9.2 update will also bump the minimum ICU to 69. If we enact a "no minimum version bump" that might prevent updating Node.js 16 to V8 9.2.

I think in that case it won't be too difficult to patch V8 to keep compatibility with earlier ICU.

> Another possibility/addition to fixing the minimum ICU was to not update ICU at all in a release line. Are there any potential downsides? We'd still be able to update timezone information (documentation on how to do so added in nodejs/node#30364).

OK for timezone information, but what about CLDR data? Could we upgrade it independently too?

**Comment by targos** [comment]
Should we try to have a special CI job that verifies that everything works with the minimum supported ICU version?

**Comment by richardlau** [comment]
> Should we try to have a special CI job that verifies that everything works with the minimum supported ICU version?

Yes. I'm looking into it.

**Comment by AdamMajer** [comment]
> > Should we try to have a special CI job that verifies that everything works with the minimum supported ICU version?
> 
> Yes. I'm looking into it.

Perhaps this CI could include other dependencies that can be linked externally too? I'm thinking especially of OpenSSL or libcares. I'm not certain about the others as they can move rather quickly and in the past Node included even unreleased versions of some (like patched version of nghttp2).

**Comment by richardlau** [comment]
We already build against shared OpenSSL (1.1.1 and 3.0.0-alpha builds) and zlib (1.2.11) (for anyone interested we do those builds in a [Docker container](https://github.com/nodejs/build/blob/master/ansible/roles/docker/templates/ubuntu1804_sharedlibs.Dockerfile.j2)).

**Comment by srl295** [comment]
> Could we patch CLDR data separately?

Short answer:  Nope! .. It would be much better to fix whatever is preventing ICU upgrades. 

> We are looking into CI for minimum ICU

Please reach out if you need help, definitely @- mention me.
